ext/datasource/refreshable_file/file_datasource.go

```diff
@@ -0,0 +1,108 @@
+package refreshable_file
+
+import (
+	"github.com/alibaba/sentinel-golang/ext/datasource"
+	"github.com/alibaba/sentinel-golang/logging"
+	"github.com/alibaba/sentinel-golang/util"
+	"github.com/fsnotify/fsnotify"
+	"github.com/pkg/errors"
+	"io/ioutil"
+	"os"
+	"time"
+)
+
+var (
+	logger = logging.GetDefaultLogger()
+)
+
+type RefreshableFileDataSource struct {
+	datasource.Base
+	sourceFilePath string
+	isInitialized  util.AtomicBool
+	watcher        *fsnotify.Watcher
+}
+
+func FileDataSourceStarter(sourceFilePath string, handlers ...datasource.PropertyHandler) (*RefreshableFileDataSource, error) {
+	var ds = &RefreshableFileDataSource{
+		sourceFilePath: sourceFilePath,
+	}
+	for _, h := range handlers {
+		ds.AddPropertyHandler(h)
+	}
+	return ds, ds.Initialize()
+}
+
+func (s *RefreshableFileDataSource) ReadSource() ([]byte, error) {
+	f, err := os.Open(s.sourceFilePath)
+	defer f.Close()
+
+	if err != nil {
+		return nil, errors.Errorf("The rules file is not existed, err: %+v.", errors.WithStack(err))
+	}
+	src, err := ioutil.ReadAll(f)
+	if err != nil {
+		return nil, errors.Errorf("Fail to read file, err: %+v.", errors.WithStack(err))
+	}
+	return src, nil
+}
+
+func (s *RefreshableFileDataSource) Initialize() error {
+	if !s.isInitialized.CompareAndSet(false, true) {
+		return errors.New("RefreshableFileDataSource had been Initialized.")
+	}
+
+	s.doUpdate()
+	w, err := fsnotify.NewWatcher()
+	if err != nil {
+		return errors.Errorf("Fail to new a w of fsnotify, err: %+v", err)
+	}
+	err = w.Add(s.sourceFilePath)
+	if err != nil {
+		return errors.Errorf("Fail add a w on file(%s), err: %+v", s.sourceFilePath, err)
+	}
+	s.watcher = w
+
+	go util.RunWithRecover(func() {
+		defer s.watcher.Close()
+		for {
+			select {
+			case ev := <-s.watcher.Events:
+				if ev.Op&fsnotify.Write == fsnotify.Write {
+					s.doUpdate()
+				}
+
+				if ev.Op&fsnotify.Remove == fsnotify.Remove || ev.Op&fsnotify.Rename == fsnotify.Rename {
+					logger.Errorf("The file source(%s) was removed or renamed.", s.sourceFilePath)
+					for _, h := range s.Handlers() {
+						err := h.Handle(nil)
+						if err != nil {
+							logger.Errorf("RefreshableFileDataSource fail to publish rules, handle: %+v; err: %+v.", h, err)
+						}
+					}
+				}
+			case err := <-s.watcher.Errors:
+				logger.Errorf("Watch err on file(%s), err: %+v", s.sourceFilePath, err)
+				time.Sleep(time.Second * 3)
+			}
+		}
+	}, logger)
+	return nil
+}
+
+func (s *RefreshableFileDataSource) doUpdate() {
+	src, err := s.ReadSource()
+	if err != nil {
+		logger.Errorf("Fail to read source, err: %+v", err)
+		return
+	}
+	for _, h := range s.Handlers() {
+		err := h.Handle(src)
+		if err != nil {
+			logger.Errorf("RefreshableFileDataSource fail to publish rules, handle: %+v; err: %+v.", h, err)
```

Comment:
(1) Here we could use the word "data" instead of `rule`. (2) Printing the handler here might not be helpful. Maybe we need to print out the type?

Reply:
(1) Right；
(2) Is type the datasource type?